File: src/main/java/ru/citeck/ecos/index/Indexes.java
package ru.citeck.ecos.index;

import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiElement;
import com.intellij.util.indexing.FileContent;
import lombok.Data;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Indexes extends HashMap<IndexKey, List<IndexValue>> {

    public FileIndexes withFileContent(FileContent fileContent) {
        return new FileIndexes((fileContent.getFile()));
    }

    @Data
    public class FileIndexes {

        private final VirtualFile virtualFile;

        public FileIndexes addSearchEverywhere(String id, PsiElement psiElement, Map<String, String> properties) {
            return add(IndexKey.SEARCH_EVERYWHERE, id, psiElement, properties);
        }

        public FileIndexes addSearchEverywhere(String id, PsiElement psiElement) {
            return add(IndexKey.SEARCH_EVERYWHERE, id, psiElement, null);
        }

        public FileIndexes addReference(String id, PsiElement psiElement, Map<String, String> properties) {
            return add(new IndexKey(IndexKey.REFERENCE_TYPE, id), id, psiElement, properties);
        }

        public FileIndexes addReference(String id, PsiElement psiElement) {
            return add(new IndexKey(IndexKey.REFERENCE_TYPE, id), id, psiElement, null);
        }

        public FileIndexes add(IndexKey key, String id, PsiElement psiElement, Map<String, String> properties) {
            return add(key, createIndex(id, psiElement, properties));
        }

        public FileIndexes add(IndexKey key, String id, PsiElement psiElement) {
            return add(key, createIndex(id, psiElement, null));
        }

        public FileIndexes add(IndexKey key, IndexValue indexValue) {
            List<IndexValue> values = get(key);
            if (values == null) {
                values = new ArrayList<>();
                put(key, values);
            }
            values.add(indexValue);
            return this;
        }

        public IndexValue createIndex(String id, PsiElement psiElement, Map<String, String> properties) {
            IndexValue indexValue = new IndexValue(id, psiElement.getTextOffset(), virtualFile.getPath());
            if (properties != null) {
                properties.forEach(indexValue::setProperty);
            }
            return indexValue;
        }

        public IndexValue createIndex(String id, PsiElement psiElement) {
            return createIndex(id, psiElement, null);
        }

    }

}

File: src/main/kotlin/ru/citeck/idea/search/index/Indexes.kt
package ru.citeck.idea.search.index

import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.PsiElement
import com.intellij.util.indexing.FileContent
import lombok.Data

class Indexes : HashMap<IndexKey, MutableList<IndexValue>>() {

    fun withFileContent(fileContent: FileContent): FileIndexes {
        return FileIndexes(fileContent.file)
    }

    inner class FileIndexes(
        private val virtualFile: VirtualFile
    ) {

        fun addSearchEverywhere(id: String, psiElement: PsiElement, properties: Map<String, String>?): FileIndexes {
            return add(IndexKey.SEARCH_EVERYWHERE, id, psiElement, properties)
        }

        fun addSearchEverywhere(id: String, psiElement: PsiElement): FileIndexes {
            return add(IndexKey.SEARCH_EVERYWHERE, id, psiElement, null)
        }

        fun addReference(id: String, psiElement: PsiElement, properties: Map<String, String>?): FileIndexes {
            return add(IndexKey(IndexKey.REFERENCE_TYPE, id), id, psiElement, properties)
        }

        fun addReference(id: String, psiElement: PsiElement): FileIndexes {
            return add(IndexKey(IndexKey.REFERENCE_TYPE, id), id, psiElement, null)
        }

        fun add(key: IndexKey, id: String, psiElement: PsiElement, properties: Map<String, String>?): FileIndexes {
            return add(key, createIndex(id, psiElement, properties))
        }

        fun add(key: IndexKey, id: String, psiElement: PsiElement): FileIndexes {
            return add(key, createIndex(id, psiElement, null))
        }

        fun add(key: IndexKey, indexValue: IndexValue): FileIndexes {
            var values: MutableList<IndexValue>? = get(key)
            if (values == null) {
                values = ArrayList()
                put(key, values)
            }
            values.add(indexValue)
            return this
        }

        fun createIndex(id: String, psiElement: PsiElement, properties: Map<String, String>?): IndexValue {
            val indexValue = IndexValue(id, psiElement.textOffset, virtualFile.path)
            properties?.forEach { (property: String, value: String) ->
                indexValue.setProperty(property, value)
            }
            return indexValue
        }

        fun createIndex(id: String, psiElement: PsiElement): IndexValue {
            return createIndex(id, psiElement, null)
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false

            other as FileIndexes

            return virtualFile == other.virtualFile
        }

        override fun hashCode(): Int {
            return virtualFile.hashCode()
        }

        override fun toString(): String {
            return "FileIndexes(virtualFile=$virtualFile)"
        }
    }
}
